Cubical/Structures/Group/Morphism.agda

```diff
@@ -0,0 +1,248 @@
+                      → (s : raw-group-structure G)
+                      → isProp (group-axioms G s)
+  isProp-group-axioms G _+_ = isPropΣ (isPropIsSemigroup _) γ
+    where
+    γ : (h : IsSemigroup _+_) →
+        isProp (Σ[ e ∈ G ] ((x : G) → (x + e ≡ x) × (e + x ≡ x))
+                         × ((x : G) → Σ[ x' ∈ G ] (x + x' ≡ e) × (x' + x ≡ e)))
+    γ h (e , P , _) (e' , Q , _) =
+      Σ≡Prop (λ x → isPropΣ (isPropΠ λ _ → isProp× (isSetG _ _) (isSetG _ _)) (β x))
+             (sym (fst (Q e)) ∙ snd (P e'))
+      where
+      isSetG : isSet G
+      isSetG = IsSemigroup.is-set h
+
+      β : (e : G) → ((x : G) → (x + e ≡ x) × (e + x ≡ x))
+        → isProp ((x : G) → Σ[ x' ∈ G ] (x + x' ≡ e) × (x' + x ≡ e))
+      β e He =
+        isPropΠ λ { x (x' , _ , P) (x'' , Q , _) →
+                Σ≡Prop (λ _ → isProp× (isSetG _ _) (isSetG _ _))
+                       (inv-lemma ℳ x x' x'' P Q) }
+        where
+          ℳ : Monoid
+          ℳ = makeMonoid e _+_ isSetG (IsSemigroup.assoc h) (λ x → He x .fst) (λ x → He x .snd)
+
+  Group→GroupΣ : Group → GroupΣ
+  Group→GroupΣ (group _ _ _ -_ isGroup) =
+   _ , _ , IsMonoid.isSemigroup (IsGroup.isMonoid isGroup) ,
+   _ , IsMonoid.identity (IsGroup.isMonoid isGroup) ,
+   λ x → (- x) , IsGroup.inverse isGroup x
+
+  GroupΣ→Group : GroupΣ → Group
+  GroupΣ→Group (_ , _ , SG , _ , H0g , w ) =
+     group _ _ _ (λ x → w x .fst) (isgroup (ismonoid SG H0g) λ x → w x .snd)
+
+  GroupIsoGroupΣ : Iso Group GroupΣ
+  GroupIsoGroupΣ = iso Group→GroupΣ GroupΣ→Group (λ _ → refl) (λ _ → refl)
+
+  group-is-SNS : SNS group-structure group-iso
+  group-is-SNS = add-axioms-SNS _ isProp-group-axioms raw-group-is-SNS
+
+  GroupΣPath : (G H : GroupΣ) → (G ≃[ group-iso ] H) ≃ (G ≡ H)
+  GroupΣPath = SIP group-is-SNS
+
+  GroupIsoΣ : (G H : Group) → Type ℓ
+  GroupIsoΣ G H = Group→GroupΣ G ≃[ group-iso ] Group→GroupΣ H
+
+  GroupIsoΣPath : {G H : Group} → Iso (GroupIso G H) (GroupIsoΣ G H)
+  fun GroupIsoΣPath (groupiso e h) = (e , h)
+  inv GroupIsoΣPath (e , h)        = groupiso e h
+  rightInv GroupIsoΣPath _         = refl
+  leftInv GroupIsoΣPath _          = refl
+
+  GroupPath : (G H : Group) → (GroupIso G H) ≃ (G ≡ H)
+  GroupPath G H =
+    GroupIso G H                    ≃⟨ isoToEquiv GroupIsoΣPath ⟩
+    GroupIsoΣ G H                   ≃⟨ GroupΣPath _ _ ⟩
+    Group→GroupΣ G ≡ Group→GroupΣ H ≃⟨ isoToEquiv (invIso (congIso GroupIsoGroupΣ)) ⟩
+    G ≡ H ■
+
+-- Extract the characterization of equality of groups
+GroupPath : (G H : Group {ℓ}) → (GroupIso G H) ≃ (G ≡ H)
+GroupPath = GroupΣ-theory.GroupPath
+
+Group-ua : {G H : Group {ℓ}} → (GroupIso G H) → (G ≡ H)
+Group-ua {G = G} {H = H} = equivFun (GroupPath G H)
+
+caracGroup-ua : {G H : Group {ℓ}} (f : GroupIso G H) → cong Group.Carrier (Group-ua f) ≡ ua (GroupIso.eq f)
+caracGroup-ua (groupiso f m) =
+  (refl ∙∙ ua f ∙∙ refl)
+    ≡⟨ sym (rUnit (ua f)) ⟩
+  ua f ∎
+
+-- Group-ua functoriality
+open Group
+
+Group≡ : (G H : Group {ℓ}) → (
```

Comment:
Is this lemma really useful?

Reply:
I'm not yet sure (apart from the proof of caracGroup≡).
I thought that some version of ΣPathP for groups might come in handy ?